text/3543-patchable-function-entry.md

```diff
@@ -0,0 +1,267 @@
+
+```
+nop
+nop
+nop
+f:
+nop
+nop
+// Code goes here
+```
+
+To set this for all functions in a crate, use `-C patchable-function-entry=nop_count=m,offset=n` where `nop_count = prefix + entry`, and `offset = prefix`. Usually, you'll want to copy this value from a corresponding `-fpatchable-function-entry=` being passed to the C compiler in your project.
+
+To set this for a specific function, use `#[patchable(prefix = m, entry = n)]` to pad with m nops before the symbol and n after the symbol, but before the prelude. This will override the flag value. To disable padding for a specific function, for example because it is part of the instrumentation framework, use `#[unpatchable]`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`patchable-function-entry` provides configurable nop padding before function symbols and after function symbols but before any generated code. We refer to the former as `prefix` padding and the latter as `entry` padding. For example, if we had a function `f` with `prefix` set to 3 and `entry` to 2, we'd expect to see:
+
+```
+f_pad:
+nop
+nop
+nop
+f:
+nop
+nop
+// Code goes here
+```
+
+Nop padding may not be supported on all architectures. As of the time of writing, support includes:
+
+- aarch64
+- aarch64\_be
+- loongarch32
+- loongarch64
+- riscv32
+- riscv64
+- i686
+- x86\_64
+
+`f_pad` addresses for every padded symbol are aggregated in the `__patchable_function_entries` section of the resulting object.
+This is not a real symbol, just a collected location.
+
+## Compiler flag `-C patchable-function-entry`
+
+This flag comes in two forms:
+
+- `-C patchable-function-entry=nop_count=m,offset=n`
+- `-C patchable-function-entry=nop_count=m`
+
+In the latter, offset is assumed to be zero. `nop_count` must be greater than or equal to `offset`, or it will be rejected.
+
+If unspecified, the current behavior is maintained, which is equivalent to `nop_count=0` here.
+
+This flag sets the default nop padding for all functions in the crate. Notably, this default *only applies to codegenned functions*. If a function is monomorphized during the compilation of another crate or any similar scenario, it will use the default from that crate's compilation. In most cases, all crates in a compilation should use the same value of `-C patchable-function-entry` to reduce confusion.
+
+`prefix` is calculated as `offset`. `entry` is calculated as `nop_count - offset`. This unusual mode of specification is intended to mimic the compiler flags of `clang` and `gcc` for ease of build system integration.
+
+Specifying the compiler flag for a backend or architecture which does not support this feature will result in an error.
```

Comment:
The GCC [docs mention](https://gcc.gnu.org/onlinedocs/gcc/Instrumentation-Options.html#:~:text=On%20PowerPC%20with%20the%20ELFv2%20ABI%2C%20for%20a%20function%20with%20dual%20entry%20points%2C%20the%20supported%20values%20for%20M%20are%200%2C%202%2C%206%20and%2014.)

> On PowerPC with the ELFv2 ABI, for a function with dual entry points, the supported values for M are 0, 2, 6 and 14.

So perhaps we should also mention that some targets may not support all values of `m` and `n` and will generate an error?

Reply:
Agreed, I'll add this in the next update.